geant4 10.7.0

drbenmorgan opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

Full disclosure - I'm a developer in Geant4, so I hope that making a version bump and slight updates to options is o.k. Apologies if not!

Other than the version bumps to the source and resources, it updates the dependencies and build options with:

- Depend on brewed `clhep` to avoid use of internal vendored version 
- Likewise, make `expat` and `zlib` `uses_from_macos` (*)
- Make the build use "global-dynamic" thread local storage. This doesn't change behaviour or dependencies, but allows the libraries to be loaded via `dlopen` etc if clients use this.

(*) I've marked as a draft because I'm having trouble getting the test to pass with the change to `uses_from_macos` for expat and zlib on my local Catalina system. The failure is one I think has been seen before, with a C++ related compile failure regarding namespaces:

```
$ brew test geant4
...
In file included from /usr/local/include/Geant4/G4Types.hh:74:
In file included from /Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/complex:245:
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/cmath:317:9: error: no member named 'signbit' in the global namespace
using ::signbit;
      ~~^
```

The same test run outside of `brew test` results in the same failure. I'm not sure if this down to Geant4's hard-coding of paths in the `Geant4PackageCache.cmake` file it installs, or to my system having both `MacOSX10.15.sdk` and `MacOSX11.0.sdk` in `/Library/Developer/CommandLineTools/SDKs`. Modifying the install to remove the hardcoded SDK paths in `Geant4PackageCache.cmake` file still gives the same error in `brew test geant4` but does resolve the issue outside of that. To try and determine whether the `brew test` error was truly related to geant4, I tried changing the `test` block to:

```ruby
  test do
    (testpath/"test.cc").write <<~EOS
      #include <complex>
      int main(){}
    EOS
    (testpath/"CMakeLists.txt").write <<~EOS
      cmake_minimum_required(VERSION 3.8)
      project(test)
      find_package(EXPAT REQUIRED)
      add_executable(test test.cc)
      target_link_libraries(test EXPAT::EXPAT)
    EOS
    system "cmake", "."
    system "make", "VERBOSE=1"
  end
```

and this reproduces the error... So I'm now at a bit of a loss to explain/fix what's going on, hence the draft to ask for better insight!

The output of `brew config` is:

```
HOMEBREW_VERSION: 2.6.0-119-g59be0ba
ORIGIN: https://github.com/Homebrew/brew
HEAD: 59be0ba88806b427b640e81ae00ccd5a00ca99bb
Last commit: 3 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 5bc442650fea9a8f958d7a254fc952891e198738
Core tap last commit: 29 minutes ago
Core tap branch: update-geant4
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: vim
HOMEBREW_FORCE_VENDOR_RUBY: set
HOMEBREW_MAKE_JOBS: 24
Homebrew Ruby: 2.6.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/bin/ruby
CPU: 24-core 64-bit ivybridge
Clang: 12.0 build 1200
Git: 2.24.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 10.15.7-x86_64
CLT: 12.2.0.0.1.1604628099
Xcode: N/A
XQuartz: 2.7.11 => /opt/X11
```

and `brew doctor` is mostly clean (the warning is I guess just to my branch for this PR):

```
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some taps are not on the default git origin branch and may not receive
updates. If this is a surprise to you, check out the default branch with:
  git -C $(brew --repo homebrew/core) checkout master
```

drbenmorgan committed:
geant4 10.7.0

- Bump version and those of resources
- Depend on brewed clhep
- Use expat and zlib from macos
- Use "global-dynamic" thread local storage to allow use via dlopen etc

carlocab commented:
Seems like a pretty common problem:

https://stackoverflow.com/questions/58628377/catalina-c-using-cmath-headers-yield-error-no-member-named-signbit-in-th

I haven't read the answers carefully enough yet to see if they present a solution to the one here, but it might be worth a read.

drbenmorgan commented:
Thanks, that's useful though not sure it provides a solution. 

The bottom line is that in Homebrew's `test` environment _on my system_ (NB: haven't diagnosed if this is down to a broken CLT install, but did reinstall from scratch yesterday), the test:

```
test do
    (testpath/"test.cc").write <<~EOS
      #include <complex>
      int main(){}
    EOS
    (testpath/"CMakeLists.txt").write <<~EOS
      cmake_minimum_required(VERSION 3.8)
      project(test)
      find_package(EXPAT REQUIRED)
      add_executable(test test.cc)
      target_link_libraries(test EXPAT::EXPAT)
    EOS
    system "cmake", "."
    system "make", "VERBOSE=1"
  end
```

will fail with signbit errors. Within the test env, the environment settings before `cmake` is invoked that might affect it are

```
$ env |sort                                                                                                                             
CC=/usr/bin/clang
CFLAGS=-Os -w -pipe -march=nehalem -mmacosx-version-min=10.15 -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk
CMAKE_FRAMEWORK_PATH=/usr/local/Frameworks:/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/System/Library/Frameworks
CMAKE_PREFIX_PATH=/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr
CPATH=/usr/local/include:/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include
CPPFLAGS=-F/usr/local/Frameworks -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk
CXX=/usr/bin/clang++
CXXFLAGS=-Os -w -pipe -march=nehalem -mmacosx-version-min=10.15 -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk
HOME=/private/tmp/geant4-test-20201207-95915-1qrgkc2
LC_CTYPE=C
LDFLAGS=-L/usr/local/lib -F/usr/local/Frameworks -Wl,-headerpad_max_install_names -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk
MAKEFLAGS=-j24
OBJC=/usr/bin/clang
OBJCFLAGS=-Os -w -pipe -march=nehalem -mmacosx-version-min=10.15 -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk
OBJCXX=/usr/bin/clang++
OBJCXXFLAGS=-Os -w -pipe -march=nehalem -mmacosx-version-min=10.15 -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk
PATH=/Library/Apple/usr/bin:/opt/X11/bin:/Applications/VMware Fusion.app/Contents/Public:/usr/local/bin:/usr/local/Homebrew/Library/Homebrew/shims/scm:/usr/bin:/bin:/usr/sbin:/sbin
PKG_CONFIG_LIBDIR=/usr/local/lib/pkgconfig:/usr/local/Homebrew/Library/Homebrew/os/mac/pkgconfig/10.15:/usr/lib/pkgconfig
```

The resultant command line that causes the failure is then:

```
/Library/Developer/CommandLineTools/usr/bin/clang++  -isystem /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include -Os -w -pipe -march=nehalem -mmacosx-version-min=10.15 -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX11.0.sdk -mmacosx-version-min=10.15 -o CMakeFiles/test.dir/test.cc.o -c /tmp/geant4-test-20201207-97111-14fb28d/test.cc
```

and it's that initial `-isystem` that causes the issue. Removing that and running the compile again results in a successful compile (the dual `isysroot` seems o.k.)

Outside of `brew test`, the same `CMakeLists.txt` and `test.cc` result in the compiler being run as:

```
/Library/Developer/CommandLineTools/usr/bin/c++   -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX11.0.sdk -mmacosx-version-min=10.15 -o CMakeFiles/test.dir/test.cc.o -c /path/to/test.cc
```

So something in the test env would appear to be causing CMake to insert that `-isystem` path when it's not needed, but it's not clear to me which env var is doing so (or if this is down to the presence of both 10.15 and 11.0 SDKs, or a broken install of these).

Also pinging @henryiii as I know he fixed/handled some of the same issues with the ROOT formula...

carlocab commented:
>So something in the test env would appear to be causing CMake to insert that -isystem path when it's not needed, but it's not clear to me which env var is doing so (or if this is down to the presence of both 10.15 and 11.0 SDKs, or a broken install of these).

I think that would be the `superenv` shims. Check `$(brew --prefix)/Homebrew/Library/Homebrew/shims/super`. I'd tag a maintainer who'd know more about them, but I think they're busy putting out fires at the moment.

`brew install --env=std` should deactivate the sanitised environment, but that doesn't help you fix this formula. I think there is a way to specify a standard build environment within the formula tho. Might help to `grep` the core repo.

henryiii commented:
This is what is in the ROOT formula:

https://github.com/Homebrew/homebrew-core/blob/1a60e2e6309cd922b8dc50cd514472daa71b5fb2/Formula/root.rb#L59-L60

Haven't checked it in a while but the comment sounds relevant.

henryiii commented on Formula/geant4.rb:
```diff
@@ -1,9 +1,9 @@
-  sha256 "ecdadbf846807af8baa071f38104fb0dcc24847c8475cd8397302e2aefa8f66f"
+  url "https://geant4-data.web.cern.ch/geant4-data/releases/source/geant4.10.07.tar.gz"
+  version "10.7.0"
+  sha256 "776ea45230d26fffebf0cf7a342af5131135759a0f70e1b4a1a401f1d1eaad4a"
```
You should add a license identifier, like `license "LGPL-2.1-or-later"` (I think there is a list of identifiers somewhere, or it's a standard, I also don't know which license Geant is under, it's one of those that doesn't self identify.

carlocab replied:
List of identifiers here: https://spdx.org/licenses/

henryiii replied:
Yep, that's it! I just don't know what one [this](https://github.com/Geant4/geant4/blob/master/LICENSE) is. :)

carlocab replied:
Well, there's always `:cannot_represent`, but that's only for if you really need it.

drbenmorgan replied:
I've added as `:cannot_represent` as there's no SPDX identifier (unsurprisingly!) and discussions here and in the main repo seemed to indicate that's a reasonable marker for cases like this?

SMillerDev replied:
That license actually looks like it's non-free and debian seems to agree: https://wiki.debian.org/DebianScience/Geant4 so that would mean we'll have to remove the formula.

carlocab replied:
Rough. Though, having formulae moved to an external tap is no longer as bad an experience for an end-user as it used to be: https://brew.sh/2020/11/18/homebrew-tap-with-bottles-uploaded-to-github-releases/

Setting up a tap for Geant4 this way should make it almost as easy to install as if it were still in Homebrew/core.

drbenmorgan replied:
Sure, can look at that now it's much easier to set up. Geant4 would still depend on `homebrew-core` packages, so if there are any tips/pointers on managing that (e.g. tracking version/revision bumps to formulae like `clhep`, `qt`), let me know.

Just to confirm for my info going forward, _any_ license with SPDX entry will be ok going forward, or must it be one from under the DFSG?

SMillerDev replied:
It's been DFSG for a while, the identifier for licenses is new though. So we don't have all formulae mapped yet.

https://docs.brew.sh/Acceptable-Formulae#we-dont-like-binary-formulae

ferdymercury replied:
Related: https://github.com/spdx/license-list-XML/issues/3038

henryiii commented on Formula/geant4.rb:
```diff
@@ -66,13 +69,13 @@ class Geant4 < Formula
+    sha256 "9954dee0012f5331267f783690e912e72db5bf52ea9babecd12ea22282176820"
   end
 
   def install
```
Can you try adding:

```ruby
# Work around "error: no member named 'signbit' in the global namespace"
    ENV.delete("SDKROOT") if DevelopmentTools.clang_build_version >= 900
```

Just to see if the same thing as ROOT works?

drbenmorgan replied:
Thanks, I've tried that with a fresh reinstall and it makes no difference to either building or installing. The formula builds/installs fine with or without this. The test still fails in both cases, and explicitly removing `SDKROOT` in the test env doesn't affect things.

There's something in the env that's set up for `test` that's confusing CMake, so I'll try and narrow this down.

carlocab replied:
Does the cmake file you use at the start of the test block come with the installation tarball, or is it something that's made during the install process?

If it's the latter, maybe it's inheriting stuff from Homebrew's build environment that makes it unsuitable for use outside it.

drbenmorgan replied:
The cmake file is from one of the example programs, so direct from the tarball without modification.

carlocab replied:
Hmmm. Odd. Ok, it may be worth trying to run the test in the install block to see if it behaves differently; maybe it's not necessarily the environment that's the culprit. (Or maybe it is and you know you just need to replicate the build environment.)

drbenmorgan committed:
geant4: add license

Marked as :cannot_represent, full version at

https://geant4.web.cern.ch/license/LICENSE.html

carlocab commented on Formula/geant4.rb:
```diff
@@ -4,6 +4,7 @@ class Geant4 < Formula
   url "https://geant4-data.web.cern.ch/geant4-data/releases/source/geant4.10.07.tar.gz"
   version "10.7.0"
   sha256 "776ea45230d26fffebf0cf7a342af5131135759a0f70e1b4a1a401f1d1eaad4a"
+  license :cannot_represent
```
I think this is fine, but can you add a comment with a link to the license just before it?

drbenmorgan replied:
Done, and I've also made sure the license file is installed (I'll patch that upstream as well)

drbenmorgan committed:
geant4: document and install license
